### PR TITLE
add support for pdfhummus

### PR DESCRIPTION
**Pull request by jingkaimori**
* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

File: packages/p/pdfhummus/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,29 @@
+-- TODO: optional compiling for tiff etc.
+
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:2]: please remove it

File: packages/p/pdfhummus/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    add_deps("freetype")
+    add_deps("libtiff")
+    add_deps("libpng")
+    add_deps("libjpeg-turbo")
+
+    on_install(function (package)
+        local configs = {"-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release")}
+        import("package.tools.cmake").install(package, configs)
+        if not package:config("shared") then 
+            os.rm(package:installdir("bin/*.dll"))
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:23]: os.tryrm

File: packages/p/pdfhummus/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        end
+    end)
+
+    on_test(function (package)
+        assert(package:has_cincludes("PDFWriter.h"))
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:28]: please use has_cfuncs to check api and links

File: packages/p/pdfhummus/xmake.lua (hunk 2 above)
Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua:23]: and .so, .dylib, .a, .lib ...

File: packages/p/pdfhummus/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,29 @@
+-- TODO: optional compiling for tiff etc.
+
+package("pdfhummus")
+    set_homepage("https://www.pdfhummus.com/")
+    set_description("High performance library for creating, modiyfing and parsing PDF files in C++ ")
+    set_license("Apache-2.0")
+
+    add_urls("https://github.com/galkahana/PDF-Write/archive/refs/tags/$(version).tar.gz",
+             "https://github.com/galkahana/PDF-Write.git")
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:9]: https://github.com/galkahana/PDF-Write 404 not found

Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua:9]: PDF-Writer

File: packages/p/pdfhummus/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    
+    add_deps("cmake")
+    add_deps("zlib")
+    add_deps("freetype")
+    add_deps("libtiff")
+    add_deps("libpng")
+    add_deps("libjpeg-turbo")
+
+    on_install(function (package)
+        local configs = {"-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release")}
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:20]: we need define `-DUSE_BUNDLED=FALSE` to disable bundled deps and find them from add_deps()

https://github.com/galkahana/PDF-Writer/blob/830dd08cfedfb67e4bbc4a95fd9e14a564882ec2/CMakeLists.txt#LL10C8-L10C19

Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua:20]: and why not add 

```
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
```

it is not supported?

File: packages/p/pdfhummus/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    on_load(function (package)
+        local function add_dep(conf, pkg)
+            if package:config(conf) then
+                package:add("deps", pkg or conf)
+            end
+        end
+        add_dep("libtiff")
+        add_dep("libpng")
+        add_dep("libjpeg","libjpeg-turbo")
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:28]: space, `, `

Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua:28]: libjpeg-turbo?

File: packages/p/pdfhummus/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,41 @@
+-- TODO: optional compiling for tiff etc.
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:1]: remove comments and empty lines

File: packages/p/pdfhummus/xmake.lua (hunk 3 above)
Comment (jingkaimori) [reply packages/p/pdfhummus/xmake.lua]: now detect main class in cxx

File: packages/p/pdfhummus/xmake.lua (hunk 5 above)
Comment (jingkaimori) [reply packages/p/pdfhummus/xmake.lua]: added deps

File: packages/p/pdfhummus/xmake.lua [hunk 8]
```diff
@@ -37,5 +37,6 @@ package("pdfhummus")
     end)
 
     on_test(function (package)
-        assert(package:has_cincludes("PDFWriter.h"))
+        assert(package:has_cxxincludes("PDFWriter.h"))
+        assert(package:has_cxxtypes("PDFWriter"))
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:40]: please use has_cfuncs or has_cxxsnippets

File: packages/p/pdfhummus/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,40 @@
+
```
Comment (waruqi) [inline packages/p/pdfhummus/xmake.lua:1]: please remove empty line

File: packages/p/pdfhummus/xmake.lua (hunk 8 above)
Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua:39]: Still not right, see 
https://github.com/xmake-io/xmake-repo/blob/ce72a2f24f83a6e064ce6e0fdadca014006eacb3/packages/l/libtins/xmake.lua#L40

File: packages/p/pdfhummus/xmake.lua (hunk 3 above)
Comment (waruqi) [reply packages/p/pdfhummus/xmake.lua]: but it will not check links

**Comment by jingkaimori** [comment]
Pdfhummus does not use `find_package` to find libaesgm, it search system install folder instead. Is it necessary to change upstream cmake file to provide xmake port?

See: https://github.com/galkahana/PDF-Writer/blob/master/CMakeLists.txt#L33

**Comment by waruqi** [comment]
you can try.

**Comment by jingkaimori** [comment]
cmakefile patched.

**Comment by waruqi** [comment]
error: ...es/private/action/require/impl/actions/patch_sources.lua:93: patch(/home/runner/work/xmake-repo/xmake-repo/packages/p/pdfhummus/patches/4.1/cmake.patch): unmatched checksum!

File: packages/p/pdfhummus/xmake.lua (hunk 8 above)
Comment (jingkaimori) [reply packages/p/pdfhummus/xmake.lua]: PDFHummus has a test binary, but it is not installed, so I include a stub test in it.

**Comment by jingkaimori** [comment]
I cannot reproduce issue of missing jpeg lib on github codespace.

**Comment by waruqi** [comment]
/usr/bin/ld: CMakeFiles/PDFWriterTestPlayground.dir/JpegLibTest.cpp.o: in function `HummusJPGSourceInitialization(jpeg_decompress_struct*, IByteReader*)':
JpegLibTest.cpp:(.text+0x94): undefined reference to `jpeg_resync_to_restart'
/usr/bin/ld: 


you can try disable tests program or pass jpeg to cmake.install(package, configs, {packagedeps = "libjpeg"})

**Comment by waruqi** [comment]
I think it would be easier for you to just use xmake.lua to port and compile it.

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake-repo/pull/1709

**Pull request closed without merging**